### PR TITLE
Return real App Catalog state and enforce actions

### LINKED ISSUE #208: Complete App Catalog: optional software UI implementation plan
## Goal

Complete App Catalog as a dependable, store-like Windows UI for Gorilla's optional software. Users should be able to find an app, understand its state, install or remove it, and see an accurate result—even after closing/reopening the UI or losing its connection to the service.

This issue is the working implementation plan. Update the stages, decisions, and linked PRs as implementation proceeds. Acceptance checkboxes describe work to complete, not claims about existing behavior.

## Agreed product scope

- **Optional software only.** Required software and general managed-update visibility are outside this release.
- Installing an optional app adds it to the device's persistent local managed installs: Gorilla should keep it installed and updated.
- Remove clears that local install selection and uninstalls once, then forgets the selection. It must not create persistent user-requested managed-uninstall policy or keep removing an app reinstalled later.
- The UI gets installed-state information from the Go service, using the same shared Go detection implementation as the CLI and scheduled runs. Do not implement separate UI detection logic.
- Item-only execution is preferred, but conditional on validating that it preserves necessary full-run lifecycle and dependency/policy behavior.
- **Store-like layout**, with prominent search and clear app status.
- A future feature may notify users when a managed update needs their action, such as closing a blocking app. That is **not a requirement for this work**.
- Preserve the existing WinUI App → portable Core → Client architecture and standard-user UI / privileged-service boundary.

Proposed supporting UX: app cards, an app detail view, contextual actions, and an activity view. Exact layouts and additional catalog presentation fields should be reviewed before implementing the visual design.

## Reviewed baseline

Repository review was performed at [5216e57481629b3252ee1beb5d13406487c9b5c0](https://github.com/1dustindavis/gorilla/commit/5216e57481629b3252ee1beb5d13406487c9b5c0). Recheck the relevant code against current main before each implementation PR.

Already present:
- Cache-first startup, live refresh, install/remove requests, operation IDs, streamed events, and post-terminal list refresh.
- Optional removal writes a persistent local managed-uninstall entry.
- MSIX packaging includes the UI and automatic LocalSystem service.
- Portable Core/Client tests, Windows integration, and source/installed-product UI validation infrastructure.

Verified gaps:
- ListOptionalInstalls currently emits placeholder metadata: item name as display name, empty version/catalog details, isInstalled=false, and status=Unknown.
- A service operation reports success when the overall managed run returns without an error. Individual installer failures can become strings ignored by the processing layer, so success does not establish the requested app's outcome.
- Progress percentages are fixed milestones around a managed run, not measured installer/download progress.
- Install and Remove are both always exposed; IsBusy is not bound to action availability.
- No client reconnect/resume workflow or durable operation recovery; service tracking is in memory.
- No manual refresh, search, app details, activity screen, or explicit loading/empty states.
- Service request validation checks command shape but does not establish optional-app eligibility.
- Earlier UI docs mix implemented work with outstanding intentions.

The [Windows UI run for the reviewed commit](https://github.com/1dustindavis/gorilla/actions/runs/34067108926) passed its real fixture install/remove checks. Its screenshots and protocol logs still show Unknown after installation. Existing E2E coverage therefore does not establish correct displayed installation state.

## Proposed stages and PR boundaries

Stages are ordered by dependency. Split a stage into focused PRs where appropriate; do not combine the entire effort into one change. Add tests and documentation with each behavior change.

### 1. Define app state, action policy, and operation results

Establish the service/Core contract before designing controls around it.

- [x] Separate observed app state, device selection/policy, and active operation state.
- [x] Define absent, installed, update available, unknown, and detection-failed behavior, including unavailable installed-version information.
- [x] Distinguish catalog/target version from observed installed version.
- [x] Define allowed actions and explain why an action is unavailable.
- [x] Define per-item results, including already satisfied/no action needed, failure, and an outcome that cannot be verified.
- [x] Define how optional requests interact with required installs, managed uninstalls, and dependencies.
- [x] Investigate the consequences of narrowing a managed run, including common setup/cleanup, reporting, shared status state, dependencies, and policy. Prefer item-only work through the serialized executor, but require evidence and regression tests before activating it.
- [x] Update protocol examples and CLI service-message expectations with the chosen contract.

Important: existing status checks answer whether an action is needed. Do not simply negate that answer to infer installed state. Script checks and missing/failed detection need explicit treatment.

**Done when:** agreed state/action tables and contract examples cover the main success, failure, conflict, and unknown-state cases, with deterministic tests for the new pure logic.

### 2. Return real optional-app information and enforce allowed actions

Depends on stage 1.

- [x] Resolve optional item names through the effective manifests and catalogs using Gorilla's existing precedence rules.
- [x] Return actual display names, available versions, relevant metadata, observed state, and allowed actions from the service. Extend the shared Go status implementation to expose its evidence while preserving CLI action-needed semantics; do not create parallel detection implementations.
- [x] Report missing/invalid catalog entries and detection problems without presenting an invented installed state.
- [x] Keep potentially slow detection from making catalog viewing unnecessarily unresponsive; make data freshness explicit.
- [x] Enforce optional-app eligibility and action policy in the service, including direct pipe/CLI requests.
- [x] Preserve the agreed persistent local managed-install behavior.
- [x] Implement one-time removal: clear local managed-install selection and request one uninstall, subject to policy. Do not persist local managed-uninstall intent; explicitly handle legacy service-generated removal entries separately from administrator policy.

**Done when:** an optional fixture displays real data and correct state; disallowed requests are rejected by the service; existing manifest/catalog and scheduled-management behavior remains covered.

### 3. Propagate accurate per-app outcomes and complete the first end-to-end flow

Depends on stages 1–2.

- [x] Carry item/action identity and structured results through installer, process, service, and UI layers.
- [x] Surface download/hash, detection, installer, script, and dependency failures at the appropriate item/operation boundary.
- [x] Do not interpret overall managed-run completion as proof of a particular app's success.
- [x] Verify post-operation state where the detection method supports it; distinguish execution failure from inability to confirm the result.
- [x] Test one-time removal followed by external reinstall and a scheduled run: the old user removal must not be enforced again.
- [x] Emit truthful queued/running/terminal states. Use indeterminate progress where measurable progress is unavailable.
- [x] Update the existing UI sufficiently to show correct state and allowed actions before the store redesign.
- [x] Extend deterministic fixtures and E2E assertions to check what the UI displays, not only on-disk markers/cache writes.

**First product milestone:** an app appears with real information, installs, displays Installed, and remains correct after reopening App Catalog. Removal has the equivalent validation. A deliberately failing fixture displays a failure rather than success.

### 4. Make tracking, duplicate handling, and recovery dependable

Depends on stage 3.

- [x] Track operations independently of page navigation and individual card instances.
- [x] Prevent duplicate/conflicting submissions in both UI and service; define behavior for multiple UI instances.
- [x] Handle an uncertain request acknowledgement without blindly resubmitting a mutation.
- [x] Provide enough service operation lookup/state to reconcile after UI reopening or pipe disconnection.
- [x] Decide and document bounded operation retention and persistence needed for service-restart recovery.
- [x] Reconcile interrupted work with observed state; do not automatically replay installers after a restart.
- [x] Show connection/status uncertainty separately from installation failure.
- [x] Preserve ongoing item state when catalog refresh replaces or updates list data.
- [x] Keep installer execution serialized unless a separately justified change establishes safe concurrency.

**Done when:** reopening the UI, dropping/reconnecting the pipe, duplicate clicks, and service interruption produce understandable states without unintended repeated actions. Cover most permutations in Core/service tests and representative real boundaries in Windows tests.

User cancellation is deferred. Canceling a UI wait must not be presented as canceling the installer.

### 5. Build the store-like catalog and app details

Depends on the state/action contract and working flow. Visual design can be reviewed earlier using representative states.

- [ ] Review a proposed layout for a searchable card grid and app detail view.
- [ ] Cards show app identity, concise descriptive information, clear status, and the appropriate action.
- [ ] Add only the presentation metadata the design needs, such as description and icon; define fallbacks for existing catalogs.
- [ ] Keep new catalog fields optional and update catalog/package-info examples, compilation, DTOs, and cache handling together.
- [ ] Search app names and supported descriptive metadata with predictable ordering and a clear no-results state.
- [ ] Details expose available version, observed state/version when known, operation feedback, and relevant action explanations.
- [ ] Keep browsing and selection stable across refreshes and operation updates.

**Done when:** real service data drives the store interface; catalogs without new optional metadata remain usable; search, details, and contextual actions work with the existing workflow.

### 6. Complete activity, refresh, and recovery presentation

Depends on stages 4–5.

- [ ] Add active and recent activity with app/action identity and accurate results.
- [ ] Show useful failure explanations and an intentional retry action when appropriate.
- [ ] Add Refresh/Retry and last-updated information.
- [ ] Separate stale/offline data warnings from app-specific action failures.
- [ ] Provide distinct loading, no software assigned, no search results, no cached data, and load-failed states.
- [ ] Keep fresh data usable if cache persistence fails; cache problems should not masquerade as service failures.
- [ ] Use plain-language errors with technical detail available for troubleshooting.

**Done when:** users can understand current work and recover from ordinary failures without restarting App Catalog or guessing whether to repeat an action.

### 7. Validate and document the complete installed product

Validation also accompanies every earlier stage; this stage closes remaining product-level gaps.

- [ ] Verify keyboard navigation, focus, accessible names/states, live announcements, scaling, resizing, contrast, and light/dark themes.
- [ ] Preserve stable automation IDs and extend them intentionally for new controls.
- [ ] Keep most behavioral permutations in portable Core/Client and Go tests.
- [ ] Extend the thin FlaUI suite for uniquely valuable full-UI workflows, retaining screenshots, automation trees, and logs.
- [ ] Validate representative supported installer/detection paths, including failures and state convergence, using reusable Windows fixtures.
- [ ] Validate the produced MSIX, standard-user UI/service communication, reopening/recovery, and package interoperability.
- [ ] Update user docs and reconcile older UI planning/architecture documents, linking this issue as the active plan.

**Done when:** the agreed optional-software experience works through the installed MSIX with passing appropriate validation and reviewed visual evidence.

## Validation contract

Reuse the Makefile and existing infrastructure from #171:
- Portable baseline: `make verify`.
- Windows service/installer/pipe changes: `make verify-windows`.
- WinUI behavior: `make verify-e2e`.
- Produced-artifact interoperability: `make verify-release GORILLA_RELEASE_EXE=<path> GORILLA_RELEASE_MSIX=<path>` on a disposable Windows machine.

Coverage and mutation results are supporting review signals, not substitutes for behavior assertions. Avoid adding a parallel test framework or percentage-driven work.

## Decisions to resolve during stage 1

The following are proposals/questions, not additional product requirements already agreed:

1. Exact precedence and explanation when optional selection conflicts with required-install or managed-uninstall policy.
2. Whether one optional request deliberately runs all pending managed changes, or has an item-specific execution scope that preserves dependency behavior.
3. Which detection methods can establish presence, installed version, and update availability reliably; how unsupported/ambiguous cases appear.
4. How to identify and recover uncertain requests and interrupted operations, including persistence and retention.
5. Whether an explicit Update action is part of this first UI or updates remain automatic with status visibility.

## References

- [Existing UI plan](https://github.com/1dustindavis/gorilla/blob/main/gorilla-ui/PLAN.md)
- [UI architecture](https://github.com/1dustindavis/gorilla/blob/main/gorilla-ui/ARCHITECTURE.md)
- [App Catalog documentation](https://github.com/1dustindavis/gorilla/blob/main/docs/app-catalog.md)
- [Validation and automation work](https://github.com/1dustindavis/gorilla/issues/171)

## Implementation PRs

Stage 1: [PR #209](https://github.com/1dustindavis/gorilla/pull/209), including the execution-scope investigation and shared Go detection boundary. Merged in `8da44b6`.

Stage 2: [PR #210](https://github.com/1dustindavis/gorilla/pull/210), returning real catalog/state data, enforcing service-owned actions, and implementing one-time removal.

Stage 3: [PR #212](https://github.com/1dustindavis/gorilla/pull/212), propagating verified per-app operation results; [PR #213](https://github.com/1dustindavis/gorilla/pull/213), consuming truthful operation results across client/Core/WinUI and validating failure and one-time-removal behavior; [PR #214](https://github.com/1dustindavis/gorilla/pull/214), validating installed/removed state across App Catalog relaunches.

Add links beneath the corresponding stages as PRs are opened. Check off acceptance criteria only after the relevant implementation and validation are complete.

### PR DESCRIPTION
**Pull request by 1dustindavis**
Stage 2 of #208 replaces App Catalog's placeholder list data and moves optional-action authorization into the privileged service.

## What changes

- Resolve effective `optional_installs` through the same ordered catalog resolver used by managed processing.
- Return real display name, catalog/version metadata, shared Go detection observations, effective policy, and service-owned allowed actions.
- Extend `pkg/status` with observation evidence while preserving `CheckStatus` action decisions, including the detection-check version authority established in stage 1.
- Reset registry evidence once per observation pass and managed run so later refreshes cannot reuse an old registry snapshot.
- Reject non-optional, policy-conflicting, unavailable, unknown, and detection-failed actions before changing local selection.
- Keep installs as persistent local managed installs.
- Make removal one-time: clear the local install selection, run one transient uninstall manifest, and remove it after the serialized managed run. Clear legacy user-generated uninstalls from `service-manifest.yaml` at service startup while preserving administrator manifests.
- Keep the v1 envelope during the transition, adding observation/policy/action objects alongside fields consumed by the current UI. Stage 3 owns the complete v2 operation/result lifecycle.
- Change the UI E2E fixture to reliable file-presence detection and assert the displayed `NotInstalled → Installed → NotInstalled` state sequence.

Optional requests still schedule the existing full managed run. Stage 3 retains the documented validation gate before switching to item-only execution.

## Validation

- `make go-format`
- Windows-targeted `make go-vet`
- race-enabled `make go-test`
- Windows test binaries compile for `pkg/service`, `pkg/status`, and `cmd/gorilla`
- `make verify` reaches the known Linux staticcheck failure for Windows-only service protocol symbols, also present on the stage 1 baseline
- Local .NET execution is unavailable in this Linux environment because the .NET CLI cannot read process start information; GitHub Windows jobs will run Client, Windows integration, and UI E2E coverage

All required GitHub checks pass on `baf8a46`: Go Tests, Verify, UI Client Tests, Windows UI Tests, and both Windows Integration jobs. The Windows UI workflow verified the real fixture's displayed `NotInstalled → Installed → NotInstalled` sequence.

This implements stage 2 only and does not close #208.

### CODE REVIEW

File: integration/windows/prepare-release-integration.ps1 [hunk 1]
```diff
@@ -479,8 +479,8 @@ $(Build-CheckScript -Path $nupkgMarker -Target "2.0.0" | ForEach-Object { "
 Ps1V1:
   display_name: Ps1V1
   check:
-    script: |
-$(Build-CheckScript -Path $ps1Marker -Target "1.0.0" | ForEach-Object { "      $_" })
+    file:
+      - path: '$ps1Marker'
```
Comment (1dustindavis) [inline integration/windows/prepare-release-integration.ps1:485]: What is the purpose of these changes?

File: pkg/process/process.go [hunk 2]
```diff
@@ -15,9 +15,9 @@ import (
 	"github.com/1dustindavis/gorilla/pkg/manifest"
 )
 
-// firstItem returns the first valid occurrence of an item in a map of catalogs.
-// It logs warnings for invalid/missing items and returns false when no valid item is found.
-func firstItem(itemName string, catalogsMap map[int]map[string]catalog.Item) (catalog.Item, bool) {
+// ResolveItem returns the first valid occurrence and its ordered catalog index.
+// Managed processing and service presentation share this precedence and validity.
```
Comment (1dustindavis) [inline pkg/process/process.go]: These comments are okay, but are a bit too terse. I don't want a book, but there is probably a better way to communicate these ideas.

File: pkg/service/run_windows.go [hunk 3]
```diff
@@ -78,6 +78,9 @@ func newServiceRunner(cfg config.Configuration, managedRun func(config.Configura
 }
 
 func (sr *serviceRunner) start(ctx context.Context) error {
+	if err := clearLegacyServiceUninstalls(sr.cfg); err != nil {
+		return fmt.Errorf("migrate service local manifest: %w", err)
```
Comment (1dustindavis) [inline pkg/service/run_windows.go]: This error message is not helpful to the person receiving it.

File: pkg/service/run_windows.go [hunk 4]
```diff
@@ -471,22 +483,43 @@ func commandFromRequestEnvelope(req serviceEnvelope[json.RawMessage]) (Command,
...
-				Status:             "Unknown",
-				StatusUpdatedAtUTC: nowRFC3339UTC(),
+				ItemName: item.ItemName, DisplayName: item.DisplayName,
+				Version: version, Catalog: item.Catalog,
+				InstallerType: detail.InstallerType, InstallerPackageID: packageID,
+				InstallerLocation: detail.InstallerLocation,
+				IsManaged:         item.Policy.Selection == appcatalog.KeepInstalled,
+				IsInstalled:       installed, Status: legacyStatus,
+				StatusUpdatedAtUTC: updated, TargetVersion: item.TargetVersion,
+				Observation: item.Observation, Policy: item.Policy, Actions: item.Actions,
```
Comment (1dustindavis) [inline pkg/service/run_windows.go]: Formatting is a bit off here. This passed go fmt?

File: pkg/status/status.go [hunk 5]
```diff
@@ -286,30 +298,67 @@ func checkAppx(catalogItem catalog.Item, installType string) (actionNeeded bool,
...
 
 // CheckStatus determines the method for checking status
 func CheckStatus(catalogItem catalog.Item, installType, cachePath string) (actionNeeded bool, checkErr error) {
+	result, err := Observe(catalogItem, installType, cachePath)
+	return result.ActionNeeded, err
+}
+
+// Observe runs the same selected check used by CheckStatus and also returns the
+// evidence needed by service consumers. Selection precedence is intentionally
+// identical to the historical CheckStatus implementation.
```
Comment (1dustindavis) [inline pkg/status/status.go:312]: I like the way this is headed. Can you breakdown the details of what is actually changing in this PR with status checks? Is observation mostly adding additional data at this point?

File: pkg/process/process.go (hunk 2 above)
Comment (1dustindavis) [reply pkg/process/process.go]: Expanded this comment. `ResolveItem` searches catalog maps in configured order, returns the first actionable installer/uninstaller definition, and returns the supplying catalog key. Managed processing and App Catalog share it so duplicate names and invalid definitions resolve identically.

File: integration/windows/prepare-release-integration.ps1 (hunk 1 above)
Comment (1dustindavis) [reply integration/windows/prepare-release-integration.ps1:485]: This release-integration fixture uses a file check so the App Catalog can verify deterministic physical presence before and after the install/remove workflow. Script requirement behavior is covered separately: when a script is selected, Gorilla keeps the existing `script -> file -> registry -> AppX` precedence and does not run the file check as supplemental evidence. I added an inline comment here and regression tests for the combined script+file case to make that intent explicit.

File: pkg/service/run_windows.go (hunk 4 above)
Comment (1dustindavis) [reply pkg/service/run_windows.go]: Reformatted the response literal and then ran the repository formatter. Confirmed on the final head: `make go-format` passes locally, and the full Go Tests workflow (including formatting, vet, staticcheck, race tests, and coverage) passes.

File: pkg/service/run_windows.go (hunk 3 above)
Comment (1dustindavis) [reply pkg/service/run_windows.go]: Updated. The error now explains that Gorilla is removing persistent uninstall requests written by an older App Catalog version, includes the exact service-manifest path, explains that startup is blocked because retaining them could repeatedly uninstall software, and preserves the underlying read/parse/write error. I also added a Windows test for the full diagnostic.

File: pkg/status/status.go (hunk 5 above)
Comment (1dustindavis) [reply pkg/status/status.go:312]: The existing `CheckStatus` action-needed result and check precedence remain unchanged. `Observe` is the richer service-facing result behind it: file/registry/AppX checks can report physical state and version evidence, while script checks keep physical state `Unknown` and expose only whether the selected script says the install requirement is `Satisfied` or `NotSatisfied`. That lets optional script items remain actionable without pretending the script proved presence. Regression tests now lock in script-first precedence even when a file check is also present.

**Comment by 1dustindavis** [comment]
Addressed the three follow-up review findings in `f8fd341`:

1. **Reconcile persisted selections with administrator policy**
   - Service-owned `managed_installs` that conflict with an administrator `managed_uninstall` are removed during catalog/action resolution and before every scheduled or requested service-managed run.
   - Unrelated local selections are preserved.
   - Added a scheduled-run regression test that verifies reconciliation occurs before managed processing.

2. **Validate dependency feasibility before authorizing Install**
   - Install authorization now walks the complete dependency graph and rejects missing or invalid dependencies, cycles, and dependencies forced absent by administrator policy.
   - Valid nested dependency graphs remain installable.
   - Added coverage confirming an unsatisfiable request is denied before any selection is persisted.
   - This does not change the legacy CLI dependency execution path.

3. **Handle ambiguous registry matches conservatively**
   - The rich `Observe` API now reports `Unknown`, omits `installedVersion`, and returns `ambiguous_registry_match` when multiple uninstall entries plausibly match.
   - Legacy `CheckStatus` keeps its existing first-match action behavior for compatibility.
   - Added deterministic ambiguity and compatibility regression tests.

Validation completed:
- `go test -race -covermode=atomic ./...`
- formatting, vet, staticcheck, and Windows service test compilation
- all five PR workflows pass: UI Client Tests, Go Tests, Windows Integration, Windows UI Tests, and Released-Binary Integration.
